packages/flutter/lib/src/widgets/router.dart

```diff
@@ -0,0 +1,1327 @@
+              .routerReportsNewRouteInformation(routeInformation);
+            _lastSeenLocation = routeInformation.location;
+          }
+        }
+        _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+        return;
+
+      case _IntentionToReportRouteInformation.must:
+        final RouteInformation routeInformation = _retrieveNewRouteInformation;
+        if (routeInformation != null) {
+          widget.routeInformationProvider
+            .routerReportsNewRouteInformation(routeInformation);
+          _lastSeenLocation = routeInformation.location;
+        }
+        _currentIntentionToReporting = _IntentionToReportRouteInformation.none;
+        return;
+    }
+  }
+
+  RouteInformation get _retrieveNewRouteInformation {
+    final T configuration = widget.routerDelegate.currentConfiguration;
+    if (configuration == null)
+      return null;
+    final RouteInformation routeInformation = widget
+      .routeInformationParser
+      .restoreRouteInformation(configuration);
+    assert((){
+      if (routeInformation == null) {
+        FlutterError.reportError(
+          const FlutterErrorDetails(
+            exception:
+              'Router.routeInformationParser returns a null RouteInformation. '
+              'If you opt for route information reporting, the '
+              'routeInformationParser must not report null for a given '
+              'configuration.'
+          ),
+        );
+      }
+      return true;
+    }());
+    return routeInformation;
+  }
+
+  void _setStateWithExplicitReportStatus(
+    _IntentionToReportRouteInformation status,
+    VoidCallback fn,
+  ) {
+    assert(status != null);
+    assert(status.index >= _IntentionToReportRouteInformation.must.index);
+    assert(() {
+      if (_currentIntentionToReporting.index >= _IntentionToReportRouteInformation.must.index &&
+          _currentIntentionToReporting != status) {
+        FlutterError.reportError(
+          const FlutterErrorDetails(
+            exception:
+              'Both Router.navigate and Router.neglect have been called in this '
+              'build cycle, and the Router cannot decide whether to report the '
+              'route information. Please make sure only one of them is called '
+              'within the same build cycle.'
+          ),
+        );
+      }
+      return true;
+    }());
+    _currentIntentionToReporting = status;
+    setState(fn);
+  }
+
+  void _maybeNeedToReportRouteInformation() {
+    _currentIntentionToReporting = _currentIntentionToReporting != _IntentionToReportRouteInformation.none
+      ? _currentIntentionToReporting
+      : _IntentionToReportRouteInformation.maybe;
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _maybeNeedToReportRouteInformation();
+  }
+
+  @override
+  void setState(VoidCallback fn) {
+    super.setState(fn);
+    _maybeNeedToReportRouteInformation();
+  }
```

Comment:
It is unusual to hock into setState like this. Can we do it the other way around? Instead of calling `setState` we call our own method (e.g. _markNeedsRebuild) or whatever, and from that method we trigger the setState and schedule the reporting? I think that makes the code a little clearer because when you see a setState call in the code you don't necessarily expect it to have the reporting side effect.